include/wil/safearrays.h

```diff
@@ -0,0 +1,1105 @@
+            {
+                vt = VT_EMPTY;  // Invalid for SAs, use to mean type couldn't be determined
+            }
+            return vt;
+        }
+
+        inline ULONG __stdcall SafeArrayGetLockCount(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            return (psa != nullptr) ? psa->cLocks : 0U;
+        }
+
+        inline HRESULT __stdcall SafeArrayCreate(VARTYPE vt, UINT dims, SAFEARRAYBOUND* sab, SAFEARRAY*& psa) WI_NOEXCEPT
+        {
+            WI_ASSERT(dims > 0);
+            psa = ::SafeArrayCreate(vt, dims, sab);
+            RETURN_LAST_ERROR_IF_NULL(psa);
+            WI_ASSERT(vt == details::SafeArrayGetVartype(psa));
+            return S_OK;
+        }
+
+        inline HRESULT __stdcall SafeArrayGetDimSize(SAFEARRAY* psa, UINT dims, ULONG* result) WI_NOEXCEPT
+        {
+            RETURN_HR_IF(E_INVALIDARG, ((dims == 0) || (dims > psa->cDims)));
+            // Read from the back of the array because SAFEARRAYs are stored with the dimensions in reverse order
+            *result = psa->rgsabound[psa->cDims - dims].cElements;
+            return S_OK;
+        }
+
+        inline HRESULT __stdcall SafeArrayGetTotalSize(SAFEARRAY* psa, ULONG* result) WI_NOEXCEPT
+        {
+            if (psa != nullptr)
+            {
+                ULONGLONG elements = 1;
+                for (UINT i = 0; i < psa->cDims; ++i)
+                {
+                    // Determine the total size of the safearray by multiplying
+                    // the sizes of each dimension together
+                    elements *= psa->rgsabound[i].cElements;
+                    if (elements > ULONG_MAX)
+                    {
+                        RETURN_HR(HRESULT_FROM_WIN32(ERROR_ARITHMETIC_OVERFLOW));
+                    }
+                }
+                *result = static_cast<ULONG>(elements);
+            }
+            else
+            {
+                // If it's invalid, it doesn't contain any elements
+                *result = 0;
+            }
+            return S_OK;
+        }
+
+        typedef resource_policy<SAFEARRAY*, decltype(&details::SafeArrayDestroy), details::SafeArrayDestroy, details::pointer_access_all> safearray_resource_policy;
+    }
+    /// @endcond
+
+    typedef unique_any<SAFEARRAY*, decltype(&details::SafeArrayUnlock), details::SafeArrayUnlock, details::pointer_access_noaddress> safearray_unlock_scope_exit;
+    typedef unique_any<SAFEARRAY*, decltype(&details::SafeArrayUnaccessData), details::SafeArrayUnaccessData, details::pointer_access_noaddress> safearray_unaccess_scope_exit;
+
+    //! Guarantees a SafeArrayUnlock call on the given object when the returned object goes out of scope
+    //! Note: call SafeArrayUnlock early with the reset() method on the returned object or abort the call with the release() method
+    WI_NODISCARD inline safearray_unlock_scope_exit SafeArrayUnlock_scope_exit(SAFEARRAY* psa) WI_NOEXCEPT

```

Comment:
Shouldn't this be named more to indicate that it's calling "lock"? E.g. `LockSafeArray_scope_exit` or something?

(I'm also not a fan of including `_scope_exit` in the name, but if that's an established naming pattern, so be it)

Reply:
I followed the naming convention used in other places, which is to name the methods and the RAII classes they return after the specific action that is performed (with `_scope_exit` appended).

Examples:
- `semaphore_release_scope_exit`  // Releases the semaphore at scope exit
- `event_reset_scope_exit`        // Invokes Reset Event at scope exit

But you're not wrong that this version is also doing a lock on construction, but at Scope Exit, it unlocks.

Reply:
> Name that function with Lock and the RAII class with unlock.

Yeah, that's basically what I was suggesting (with an additional side note that the trailing `_scope_exit` in the function name is excessive). I see that some functions have it, but some others don't (e.g. `wil::CoInitializeEx` etc.)